Makefile.cache

```diff
@@ -0,0 +1,637 @@
+	$(call GET_MOD_SHA,$(1))
+
+	# Form the cache file name
+    $(eval $(1)_MOD_CACHE_FILE      := $(1)-$($(1)_DEP_MOD_SHA)-$($(1)_MOD_HASH).tgz)
+	$(if $(MDEBUG), $(info MODE_CACHE_FILE := $($(1)_MOD_CACHE_FILE)))
+
+	# Retrive and log files list that are modified for the target.
+	$(call GET_MODIFIED_FILES,$(1))
+	$(if $($(1)_FILES_MODIFIED),
+		echo "Target $(1) dependencies are modifed - global cache skipped" >> $($(1)_DST_PATH)/$(1).log
+		echo "Modified dependencies are : [$($(1)_FILES_MODIFIED)] " >> $($(1)_DST_PATH)/$(1).log
+		$(eval $(1)_CACHE_DIR := $(SONIC_DPKG_LOCAL_CACHE_DIR)))
+
+	# Choose the cache file path in the following order
+	# 1. First load from Local cache path
+	# 2. If not, load from global cache path
+	$(eval CACHE_FILE_SELECT:=$(or $(wildcard $(SONIC_DPKG_LOCAL_CACHE_DIR)/$($(1)_MOD_CACHE_FILE)), \
+                                   $(wildcard $(SONIC_DPKG_CACHE_DIR)/$($(1)_MOD_CACHE_FILE))) )
+
+	# Check if any of the derived package is not built
+	$(eval LOAD_DRV_DEB := $(foreach pkg,$(addprefix $($(1)_DST_PATH)/,$(1) $($(1)_DERIVED_DEBS) $($(1)_EXTRA_DEBS)),$(if $(wildcard $(pkg)),,$(pkg))))
+
+	# Load the cache if cache is eanbled and cache file is present in the cache
+	# Update the cache_loaded variable
+	$(if $(and $(CACHE_FILE_SELECT),$(filter cache,$(SONIC_DPKG_CACHE_METHOD))),
+		$(if $(LOAD_DRV_DEB), $($(1)_CACHE_USER) tar -C $($(1)_BASE_PATH) -mxzvf $(CACHE_FILE_SELECT) 1>> $($(1)_DST_PATH)/$(1).log ,echo );
+		echo "File $(CACHE_FILE_SELECT) is loaded from cache" >> $($(1)_DST_PATH)/$(1).log
+		$(eval $(1)_CACHE_LOADED := Yes)
+		$(shell touch $(CACHE_FILE_SELECT))
+		echo "[ CACHE::LOADED ] $($(1)_CACHE_DIR)/$($(1)_MOD_CACHE_FILE)" >> $($(1)_DST_PATH)/$(1).log
+		,
+		echo "File $($(1)_CACHE_DIR)/$($(1)_MOD_CACHE_FILE)  is not present in cache or cache mode set as $(SONIC_DPKG_CACHE_METHOD) !" >> $($(1)_DST_PATH)/$(1).log
+		echo "[ CACHE::SKIPPED ] $($(1)_CACHE_DIR)/$($(1)_MOD_CACHE_FILE)" >> $($(1)_DST_PATH)/$(1).log
+		echo "[ CACHE::SKIPPED ] DEP_FILES - Modified Files: [$($(1)_FILES_MODIFIED)] " >> $($(1)_DST_PATH)/$(1).log
+		echo "[ CACHE::SKIPPED ] DEPENDS   - Modified Files: [$?] " >> $($(1)_DST_PATH)/$(1).log
+	 )
+endef
+
+# Saves the deb package into debian cache
+# A single tared-zip cache is created for .deb and its derived packages in the cache direcory.
+# It saves the .deb into global cache only when its dependencies are not changed,
+#    Otherwise it saves the .deb into local cache
+# The cache save is protected with lock.
+# The SAVE_INTO_CACHE macro has dependecy with LOAD_FROM_CACHE macro
+#      The target specific variables -_MOD_SRC_PATH, _MOD_CACHE_FILE and _FILES_MODIFIED are
+#      derived from the LOAD_FROM_CACHE macro
+# Args:
+# $(1) => target name
+# $(2) => target output file name
+define SAVE_INTO_CACHE
+	$(eval MOD_CACHE_FILE=$($(1)_MOD_CACHE_FILE))
+	$(call MOD_LOCK,$(1),$(SONIC_DPKG_CACHE_DIR),$(MOD_CACHE_LOCK_SUFFIX),$(MOD_CACHE_LOCK_TIMEOUT))
```

Comment:
When local cache is used, why add .lock in the global cache folder?
xumia@SonicVM1:~$ ll /nfs/dpkg_cache/ | grep header
-rwxrwxrwx 1 xumia   xumia          0 Feb 15 12:08 linux-headers-4.9.0-9-2-common_4.9.168-1+deb9u5_all.deb_cache_access.lock*
xumia@SonicVM1:~$ ll /projects/xumia/test/sonic-buildimage/target/cache/ | grep header
-rwxrwxrwx 1 xumia xumia  47413370 Feb 15 12:08 linux-headers-4.9.0-9-2-common_4.9.168-1+deb9u5_all.deb-70715c539771dc4462e97a9-269d4dcac49c5d202ec07c4.tgz*

Reply:
Agreed.   We will address this in next phase.
The sonic build uses hierarchy of cache in the order of 1. local, 2. global.  SInce sonic build jobs runs in parallel, so its better to have global lock even for local cache and this would simplify the CACHE_LOAD macro instead of having lock only for global cache. 

Some module has a special characters in the target filename which causes the flock to fail. so we removed those special characters.